keys.go

```diff
@@ -0,0 +1,70 @@
+package paseto
+
+// keyPurpose indicates if key is symmetric, private or public
+type keyPurpose int
+
+// keyVersion indicates the token version the key may be used for
+type KeyVersion int
+
+const (
+	// Invalid key version
+	KeyVersionInvalid KeyVersion = 0
+	// Key used for V1 tokens
+	KeyVersionV1 KeyVersion = 1
+	// Key used for V2 tokens
+	KeyVersionV2 KeyVersion = 2
+	// Key used for V3 tokens
+	KeyVersionV3 KeyVersion = 3
+	// Key used for V4 tokens
+	KeyVersionV4 KeyVersion = 4
+
+	// Symmetric key used for local tokens
+	keyPurposeLocal keyPurpose = 1
+	// Asymmetric secret key used for public tokens
+	keyPurposeSecret keyPurpose = 2
+	// Asymmetric public key used for public tokens
+	keyPurposePublic keyPurpose = 3
+)
+
+type Key interface {
+	// Export raw key data as hex string
+	ExportHex() string
+	// Export raw key data as byte array
+	ExportBytes() []byte
+	// Returns purpose of key
+	getPurpose() keyPurpose
+	// Returns the version of the paseto tokens the key may be used for
+	getVersion() KeyVersion
+}
+
+// Convert key version to PASERK header string
+func KeyVersionToString(version KeyVersion) string {
+	switch version {
+	case KeyVersionV1:
+		return "k1"
+	case KeyVersionV2:
+		return "k2"
+	case KeyVersionV3:
+		return "k3"
+	case KeyVersionV4:
+		return "k4"
+	default:
+		return ""
+	}
+}
+
+// Parse key version from PASERK header string (eg. "k2")
+func KeyVersionFromString(versionStr string) KeyVersion {
+	switch versionStr {
+	case "k1":
+		return KeyVersionV1
+	case "k2":
+		return KeyVersionV2
+	case "k3":
+		return KeyVersionV3
+	case "k4":
+		return KeyVersionV4
+	default:
+		return KeyVersionInvalid
+	}
+}
```

Comment:
My preference for this would be to return `(KeyVersion, error)`, and then have the `default:` case return an error (rather than introducing a `KeyVersionInvalid` value into `KeyVersion`. And then similar comment for the similar bits of code, where an `*Invalid` variant is introduced.

Reply:
Hi, it is quite simple to add a second return value with error type, but even that case something must be returned as KeyVersion as well. The following statement would be invalid:
```
default:
    return nil, fmt.Errorf("invalid key version string: %s", versionStr)
```
The return value must be an integer (which is the underlying type of KeyVersion). Default value of int is 0, which is the preferred return value on error.
So I could use: 
```
default:
    return 0, fmt.Errorf("invalid key version string: %s", versionStr)
```
or equivalently (as `const KeyVersionInvalid KeyVersion = 0`

```
default:
    return KeyVersionInvalid, fmt.Errorf("invalid key version string: %s", versionStr)
```

So if you think, it is cleaner to return an error object as well, I could simply add it, just please let me know. I'm not very familiar with coding conventions in go.

Reply:
Ah you're right, sorry I overlooked that. 

Do we actually need this pattern at all if we set:

```go
const KeyVersionV1 = "k1"
```

etc...?

If we do, I'd still prefer we return an error than rely on callers checking for the invalid value, as the error is more explicit from a function signature standpoint, so harder to forget. Either we can assign the const to a locally scoped var before returning (if using pointers), or return the invalid zero value (though maybe we don't need to actually export that invalid value const as part of the API).